Allow underscores in numbers. Closes #1677

Balajiganapathi opened the pull request:
Closes #1677.

Follows https://www.python.org/dev/peps/pep-0515/

axic commented on libsolidity/parsing/Scanner.cpp:
```diff
@@ -753,10 +773,22 @@ Token::Value Scanner::scanNumber(char _charSeen)
 				if (!isHexDigit(m_char))
 					return Token::Illegal; // we must have at least one hex digit after 'x'/'X'
-				while (isHexDigit(m_char))
+				while (isHexDigit(m_char) || m_char == '_') // same logic as scanDecimalDigits
```
This doesn't prevent the leading underscore which seems to be weird in hex (`0x_1234`)

Balajiganapathi replied:
I added that as python allows it:
 (an example from [https://www.python.org/dev/peps/pep-0515/](https://www.python.org/dev/peps/pep-0515/))
flags = 0b_0011_1111_0100_1110

axic commented:
Thanks! Can you please add tests that (assuming we strictly follow PEP-0515):
- trailing underscores are not allowed (in both decimal, hex and exponential)
- underscores within the exponent part is allowed (`1e3_00_0`)
- leading underscores in decimals (disallowed by PEP)
- consecutive underscores are not allowed in any of the places where underscores are allowed

Deviating from the PEP I'd argue that leading underscores in hex are a nuisance.

Also please update the documentation (search for "rational and integer literals").

axic commented:
Also there's this rule in the PEP:
> For the b, x and o format specifiers, _ will be allowed and group by 4 digits.

e.g. `0x1_2_3` is not valid, but `0x1234_5678` is

Balajiganapathi commented:
I agree with you that 0x_abc looks weird and not consistent with disallowing leading underscore for decimal numbers.

> Also there's this rule in the PEP:
> For the b, x and o format specifiers, _ will be allowed and group by 4 digits.
e.g. 0x1_2_3 is not valid, but 0x1234_5678 is

My understanding is that this is for the _output_ format specifier, when converting from number to string. This is not a constraint on the syntax of a numeric literal. The [grammar](https://www.python.org/dev/peps/pep-0515/#literal-grammar) given does not specify this constraint.

I am working on adding the tests and disallowing leading underscore. Is this additional grouping constraint really needed?

Balajiganapathi commented:
@axic  I have added more tests and docs.

I also tried in python3.6 and it does allow arbitrary grouping of digits.

e.g. `x = 0xab_cd_e_f` is valid

pirapira commented:
@axic do you still want precisely sized groupings?

axic commented:
Even the PEP introductory section suggests that. I think it definitely is a better idea to enforce that (we can always lax the rules, but harder to make them strict once out in a lax version).

Balajiganapathi commented:
@axic  Ok will do this then. 

Just to be clear the following will be invalid:

0x_abcd_cdef (leading underscore)
0xab_cd (not grouped by 4)

What about the following?
0x12345678_9abc - one group of 4 has underscore but other doesn't.
0x123_4567 - The total number of digits is not divisible by 4.

axic commented:
Good questions!

> 0x12345678_9abc - one group of 4 has underscore but other doesn't.

I guess this is fine too (I would personally not allow it, but I guess could be used to show where the fixed point would be after conversion? Tiny use case though). Perhaps this is the point where lax rules gain ground.

> 0x123_4567 - The total number of digits is not divisible by 4.

I think this is fine.

Balajiganapathi commented:
@axic Thanks One more question :)

Since we are allowing 0x123_4567, should we allow 0x1234_567 too? I think allowing only one of these 2 makes sense (and will make implementation slightly easier).

axic commented:
I wouldn't allow `0x1234_567`.

Balajiganapathi commented:
Hi, I am thinking about the best way to implement this. I think this rule will cover all the cases. Or do you want more restrictive ones?

* The number of digits to the *right* of each underscore should be a multiple of 4. 
* And ofcourse, no leading, trailing or double underscores

This will take care of all the above cases. But it will also allow such cases:
`0x12341234_123456781234_abcd`

chriseth commented:
Due to our ambiguous handling of literals, we should allow both `0x123_4567` and `0x1234_567`. The first makes sense if used in a number context, the second if used in a `bytes32` context.

Ok, perhaps to complicate things even further:

If the literal ends in 4 hex digits without separator, the length of the first element is irrelevant. If it starts with four hex digits without separator, the last element has to have an even length. So `0x1234_567` would be disallowed, but `0x1234_5670` would not be.

axic commented:
I'd argue that in a number context one could say leaving the leading 0 nibble out is acceptable (as it is a number), while in a `bytes32` context it isn't since that refers to actual bytes.

`0x1234_567` is very misleading, because even in a `bytes32` context it will be assigned as `0x01234567...0`.

chriseth commented:
Exactly! Although the "context" should not influence whether the literal is valid or not, it should be visible from looking at the literal itself.

axic commented:
I can't follow you, you say `0x1234_567` should be allowed because it makes sense for `bytes32`. I do not think it makes sense for it at all.

Looking at `bytes32 x = 0x1234_567` I'd get the impression that it will be `0x12345670...0`, but it will be `0x01234567...0`.

chriseth commented:
Ok, sorry, let me clarify: `0x1234_567` should not be allowed, but `0x123_4567` should be.

axic commented:
Also I'd say the parser should not enforce any rules (perhaps apart from stray leading, trailing underscores), but rather have it in the SyntaxChecker.

Reason:
- better error reporting
- perhaps less code (could use a regular expression there).
- and easier to describe it in `grammar.txt`

Balajiganapathi commented:
@chriseth  @axic , is the following rule fine?

* If there is a semicolon in hex numeric literal, then it should be before every 4th digit from right. 

> but rather have it in the SyntaxChecker.

How about in RationalNumberType::isValidLiteral in Types.cpp file?

 > and easier to describe it in grammar.txt

I could not find this file in this repo :(

Balajiganapathi commented:
Also I am assuming this does not apply to base 10 numerals as the rules of separating are different in different countries.

chriseth commented:
@axic great idea about syntaxchecker! This would also automatically allow the parsing to continue in case of an error.

@Balajiganapathi I would prefer the following rule:

If there is no separator at all, it is fine.
If there is a separator, the number of hex digits between separators has to be exactly 4 and special rules about the part before the first separator and after the last one are:

If the literal ends in 4 hex digits without separator, the length of the first element can be arbitrary. If it starts with four hex digits without separator, the last element has to have an even length. If neither the first nor the last has 4 hex digits without separator, the literal is invalid.

Balajiganapathi commented:
Not able to find out why ci build is failing. It passed in first commit. In next 2 commits I only made some cosmetic changes, yet the ci is showing build failed. I looked at build logs, it looks like it is randomly failing some.

axic commented:
```
Downloading version soljson-v0.4.18+commit.9cf6e910.js
Hash mismatch: 0x0478b43de978b1af1d6d6d8c09e84cdb2cc8ed76218d38f17b841b6e539742f0 vs 0xcecc3e8b4d1a9bb6ceadae7f681def7982755cbe97a4d28c043d3136ccbe7df1
```

This failure is down to Travis, you cannot do much against it.

Balajiganapathi commented:
@axic / @chriseth  can you please review the changes.

chriseth commented on libsolidity/analysis/SyntaxChecker.cpp:
```diff
@@ -180,6 +182,39 @@ bool SyntaxChecker::visit(Throw const& _throwStatement)
+	if (!_literal.isHexNumber())
+		return true;
+	// We have a hex literal. Do underscore validation
+	ASTString value = _literal.value().substr(2); // Skip the 0x
```
Please add an assert that the first two characters are `0x`.

chriseth commented on libsolidity/analysis/SyntaxChecker.cpp:
```diff
@@ -180,6 +182,39 @@ bool SyntaxChecker::visit(Throw const& _throwStatement)
+	for (size_t i = 1; i + 1 < parts.size(); ++i)
+	{
+		if (parts[i].size() != 4)
+			m_errorReporter.syntaxError(_literal.location(), "Invalid use of underscores in hex literal");
```
Could you make the error message a little more specific? Something like
`Invalid use of underscores in hex literal. Found inner part of length 2 (has to be 4 characters).`

chriseth commented on libsolidity/analysis/SyntaxChecker.cpp:
```diff
@@ -180,6 +182,39 @@ bool SyntaxChecker::visit(Throw const& _throwStatement)
+		return true;
+
+	// Both first and last part is invalid
+	m_errorReporter.syntaxError(_literal.location(), "Invalid use of underscores in hex literal");
```
Please also make this error message more specific.

chriseth commented on libsolidity/parsing/Scanner.cpp:
```diff
@@ -755,8 +768,21 @@ Token::Value Scanner::scanNumber(char _charSeen)
+						advance();
+						if (!isHexDigit(m_char))
+						{
+							rollback(1);
```
I think control flow would be clearer if you move the `rollback(1)` before the if statement (and of course save `m_char` in a local variable).

chriseth replied:
Ok, reading that again - why do you advance if you rollback later anyway?

Balajiganapathi commented:
@chriseth  I have added the assert + detailed error messages.

I have also refactored the scanning code so that there is no need for weird rollbacks. Hope it is clearer now.

chriseth commented:
Sorry to nitpick, but somehow I have the feeling that the parsing logic can be even more simplified.
We try to model the regular expression `[:digit:]+(_[:digit:]+)*` right?
Shouldn't this be possible with something like

```
do
{
  if (!isDecimalDigit(m_char)) return;
  while (isDecimalDigit(m_char)) advance();
  if (m_char == '_')
  {
   advance();
   if (!isDecimalDigit(m_char)) { rollback(1); return; }
  }
}
while (isDecimalDigit(m_char));
```

axic commented:
Can the number of commits be squashed a bit?

axic commented:
Also could the inner parsing made into a template to be used by both hex and decimal?

axic commented on libsolidity/analysis/SyntaxChecker.cpp:
```diff
@@ -180,6 +183,43 @@ bool SyntaxChecker::visit(Throw const& _throwStatement)
+{
+	if (!_literal.isHexNumber())
+		return true;
+	// We have a hex literal. Do underscore validation
```
Why isn't this code also validating for decimals?

Balajiganapathi replied:
@axic For integers, fractional and scientific number with decimal digits, scanner itself takes care of underscores. For hex, SyntaxChecker validates. We could move both to SyntaxChecker I guess, but underscore validation for numbers with . and scientific notation would be complicated.

Balajiganapathi commented:
@chriseth Yes that is a much better way to implement the regex. I have used that snippet now in scanDecimalDigits.

@axic Sure will squash some commits.

Balajiganapathi commented:
@chriseth I have made the changes.

chriseth commented on libsolidity/parsing/Scanner.cpp:
```diff
@@ -755,8 +772,18 @@ Token::Value Scanner::scanNumber(char _charSeen)
+				while (isHexDigit(m_char) || m_char == '_') // Unlike decimal digits, we keep the underscores for later validation
+				{
+					if (m_char == '_') {
+						if (last == '_')
```
This will always be triggered for a single `_`.

Balajiganapathi replied:
@chriseth  I didn't understand what you meant.

The idea is this: Line 773 invalidates leading underscore. After that, we only need to avoid double undersocre. So, we store the last character. If current char is an underscore and last char was also '_' then it is invalid. Otherwise we can add the current character and validate later in SyntaxChecker

chriseth replied:
Ah right, now I see it, sorry. If you want, you can contract the two if statements. Otherwise, at least move the `{` to the next line :-)

Balajiganapathi replied:
@chriseth Done. Too many formatting issues, sorry :( .

chriseth commented on libsolidity/parsing/Scanner.cpp:
```diff
@@ -726,8 +726,25 @@ Token::Value Scanner::scanHexString()
-		addLiteralCharAndAdvance();
+	// Parse for regex [:digit:]+(_[:digit:]+)*
+
+	do {
```
The `{` should go to the next line.

chriseth reviewed (dismissed):
Fine with me. @axic?

axic commented on test/libsolidity/SolidityNameAndTypeResolution.cpp:
```diff
@@ -7313,6 +7313,59 @@ BOOST_AUTO_TEST_CASE(no_address_members_on_contract)
+				uint256 a = 0x1234_abcd;
+				uint256 b = 0x1234_5678_9abc;
+				uint256 c = 0x123_4567;
+				uint256 d = 0x1234_56 wei;
```
Hm, I wonder if we should actually disallow this in the next breaking release :)

axic replied:
These tests need to be updated after #3628.

Balajiganapathi replied:
@axic So will have to remove tests that includes an unit after the hex literal correct?

axic commented on test/libsolidity/SolidityNameAndTypeResolution.cpp:
```diff
@@ -7313,6 +7313,59 @@ BOOST_AUTO_TEST_CASE(no_address_members_on_contract)
+				uint256 d = 0x1234_56 wei;
+				uint256 e = 0x12345678_9abc_abcd_abcd;
+                uint256 f = 0x1234_5678_9abc_abcdabcd;
+                uint256 x = 0x123456789abcdef;
```
Indentation.

axic commented on test/libsolidity/SolidityNameAndTypeResolution.cpp:
```diff
@@ -7313,6 +7313,59 @@ BOOST_AUTO_TEST_CASE(no_address_members_on_contract)
+			}
+		}
+	)";
+	CHECK_ERROR(text, SyntaxError, "Invalid use of underscores in hex literal. Found inner part of length 8 (has to be 4 digits).");
```
Uses `length` and `digits`. I think it should just use one term and `digits` seems to be fine. This applies to the other messages too.

axic replied:
@chriseth this hasn't been addressed

pirapira commented:
Please rebase.

Balajiganapathi commented:
@pirapira Rebased.

chriseth commented:
@axic please merge if you are fine with it.

pirapira commented on Changelog.md:
```diff
@@ -16,6 +16,7 @@ Features:
  * Syntax Checker: Turn the usage of ``callcode`` into an error as experimental 0.5.0 feature.
  * Type Checker: Improve address checksum warning.
  * Type Checker: More detailed errors for invalid array lengths (such as division by zero).
+ * Allow underscores in numeric literals
```
Please move this entry to `0.4.20`.

chriseth commented:
Rebased.

chriseth commented:
Moved to 0.5.0

chriseth commented:
@ekpyron @bit-shift does one of you want to continue this? There are still some unaddressed comments by @axic.

Balajiganapathi commented:
@chriseth I will fix those this weekend.

axic commented:
Rebased.

christianparpart commented:
not sure how I closed this PR?
